### ISSUE TITLE
Valid country code refused

### ISSUE THREAD

**Issue by remifleurance** [idev1/rpihotspot#12]
Using the following command:
```
sudo ./setup-network.sh --install --ap-ssid="<ssid>" --ap-password="<pass>" --ap-country-code="FR"
```
The script refuses the country code, which is valid:
```
Invalid Access Point(AP) setup options are specified for installation.
Please provide the below [OPTION] for installation:
============================================================================



--ap-country-code      Optional field for installation: Set Access Point(AP) Country Code. Default value is: IN.
                       Make sure that  the entered Country Code matches WiFi Country Code if it exists in /etc/wpa_supplicant/wpa_supplicant.conf
                       Allowed Country codes are:
                       AD, AE, AF, AG, AI, AL, AM, AO, AQ, AR, AS, AT, AU, AW, AX, AZ, BA, BB, BD, BE, BF, BG, BH, BI, BJ, BL, BM, BN, BO, BQ,
                       BR, BS, BT, BV, BW, BY, BZ, CA, CC, CD, CF, CG, CH, CI, CK, CL, CM, CN, CO, CR, CU, CV, CW, CX, CY, CZ, DE, DJ, DK, DM,
                       DO, DZ, EC, EE, EG, EH, ER, ES, ET, FI, FJ, FK, FM, FO, FR, GA, GB, GD, GE, GF, GG, GH, GI, GL, GM, GN, GP, GQ, GR, GS,
                       GT, GU, GW, GY, HK, HM, HN, HR, HT, HU, ID, IE, IL, IM, IN, IO, IQ, IR, IS, IT, JE, JM, JO, JP, KE, KG, KH, KI, KM, KN,
                       KP, KR, KW, KY, KZ, LA, LB, LC, LI, LK, LR, LS, LT, LU, LV, LY, MA, MC, MD, ME, MF, MG, MH, MK, ML, MM, MN, MO, MP, MQ,
                       MR, MS, MT, MU, MV, MW, MX, MY, MZ, NA, NC, NE, NF, NG, NI, NL, NO, NP, NR, NU, NZ, OM, PA, PE, PF, PG, PH, PK, PL, PM,
                       PN, PR, PS, PT, PW, PY, QA, RE, RO, RS, RU, RW, SA, SB, SC, SD, SE, SG, SH, SI, SJ, SK, SL, SM, SN, SO, SR, SS, ST, SV,
                       SX, SY, SZ, TC, TD, TF, TG, TH, TJ, TK, TL, TM, TN, TO, TR, TT, TV, TW, TZ, UA, UG, UM, US, UY, UZ, VA, VC, VE, VG, VI,
                       VN, VU, WF, WS, YE, YT, ZA, ZM, ZW

```

I check the script, this value is compared to the one in `wpa_supplicant.conf`. But it is actually the same:
```
sudo cat /etc/wpa_supplicant/wpa_supplicant.conf | grep 'country=' | awk -F '=' '{print $2}'
```
```
FR
```

Notes:
- The Pi's network was configured with the "headless" method (by creating wpa_supplicant.conf on the boot partition)
- This is a fresh install of Raspbian Buster
- I tried with other codes, same issue

Edit : the issue also occurs on Raspbian Stretch

**Comment by idev1** [comment]
> Hi, thanks for using the script and I appreciate you for reporting the issues !

I replicated the reported issue by following the below given steps. 

1. Downloaded latest Raspbian lite Buster OS using link: https://downloads.raspberrypi.org/raspbian_lite_latest

2. Written the above downloaded image to SD card (I have used 8GB Class 4 SD Card).

3. Created a new file: "wpa_supplicant.conf" under "boot" folder of SD card and updated the below given contents in the same file. (Note: I referred the link: "https://desertbot.io/blog/headless-raspberry-pi-3-bplus-ssh-wifi-setup" - step 4). 

```
country=IN
update_config=1
ctrl_interface=/var/run/wpa_supplicant

network={
  scan_ssid=1
  ssid="xxxxx"
  psk="xxxxx"
}
```

4. Inserted the freshly created SD card (by following all the above given steps) into my Raspberry Pi 3B+ device and booted my Pi. Here, I changed my Raspberry Pi initial configuration to enable SSH, changed Keyboard Layout etc using command: "sudo raspi-config".

5. After boot, updated the OS using below given commands:

```
sudo apt update
sudo apt upgrade
```

6. Installed "Git" using below given command:
```
sudo apt install git
```

7. Cloned my Github script using below given command:
```
git clone https://github.com/idev1/rpihotspot.git
```

8. Step 7 will create a folder "rpihotspot". Enter into the directory.
```
cd rpihotspot
```

9. Given execute permission to the script: "setup-network.sh" using the below command.
```
sudo chmod ug+x ./setup-network.sh
```

10. Now, execute the script: "setup-network.sh" using the below given command (note: you can change the command as per your need).
```
sudo ./setup-network.sh --install --ap-ssid="xxxxx" --ap-password="xxxxx" --ap-country-code="IN"
```

11. The above step execution will produce the sample logs (given below) and if everything goes successful and after finishing the script, after a predefined wait period of 10 seconds, the system will automatically reboot.

```
[WLAN]: wlan0 IP address: 192.168.43.233
[WLAN]: wlan0 IP Mask address: 255.255.255.0
[WLAN]: wlan0 IP Broadcast address: 192.168.43.255
[WLAN]: wlan0 Country Code: IN
[WLAN]: wlan0 Channel: 1
[Cleanup]: cleaning ...
[Cleanup]: Cleaned all NAT IP Table entries.
Reading package lists... Done
Building dependency tree
Reading state information... Done
0 upgraded, 0 newly installed, 0 to remove and 20 not upgraded.
[Cleanup]: DONE
[Install]: installing: hostapd dnsmasq iptables-persistent from net ...
Hit:1 http://archive.raspberrypi.org/debian buster InRelease
Hit:2 http://raspbian.raspberrypi.org/raspbian buster InRelease
Preconfiguring packages ...
Selecting previously unselected package hostapd.
(Reading database ... 40782 files and directories currently installed.)
Preparing to unpack .../0-hostapd_2%3a2.7+git20190128+0c1e29f-6+deb10u1_armhf.deb ...
Unpacking hostapd (2:2.7+git20190128+0c1e29f-6+deb10u1) ...
Selecting previously unselected package netfilter-persistent.
Preparing to unpack .../1-netfilter-persistent_1.0.11_all.deb ...
Unpacking netfilter-persistent (1.0.11) ...
Selecting previously unselected package iptables-persistent.
Preparing to unpack .../2-iptables-persistent_1.0.11_all.deb ...
Unpacking iptables-persistent (1.0.11) ...
Selecting previously unselected package dns-root-data.
Preparing to unpack .../3-dns-root-data_2019031302_all.deb ...
Unpacking dns-root-data (2019031302) ...
Selecting previously unselected package dnsmasq-base.
Preparing to unpack .../4-dnsmasq-base_2.80-1+rpt1_armhf.deb ...
Unpacking dnsmasq-base (2.80-1+rpt1) ...
Selecting previously unselected package dnsmasq.
Preparing to unpack .../5-dnsmasq_2.80-1+rpt1_all.deb ...
Unpacking dnsmasq (2.80-1+rpt1) ...
Setting up dnsmasq-base (2.80-1+rpt1) ...
Setting up dns-root-data (2019031302) ...
Setting up dnsmasq (2.80-1+rpt1) ...
Created symlink /etc/systemd/system/multi-user.target.wants/dnsmasq.service â /lib/systemd/system/dnsmasq.service.
Setting up hostapd (2:2.7+git20190128+0c1e29f-6+deb10u1) ...
...
[Install]: enabling netStop.service ...
...
[Install]: DONE
[Reboot]: In 10 seconds ...
```

12. After reboot, I checked to see if the Access Point (AP) setup is correctly done or not using the below given command.
```
ifconfig
```

13. Above execution of command is producing the below given information to me.
```
eth0: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        ether b8:27:eb:56:50:97  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

uap0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 10.0.0.1  netmask 255.0.0.0  broadcast 10.255.255.255
        inet6 fe80::ba27:ebff:fe03:5c2  prefixlen 64  scopeid 0x20<link>
        ether b8:27:eb:03:05:c2  txqueuelen 1000  (Ethernet)
        RX packets 4624  bytes 888091 (867.2 KiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 5056  bytes 3541946 (3.3 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

wlan0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 1500
        inet 192.168.43.233  netmask 255.255.255.0  broadcast 192.168.43.255
        inet6 fe80::6d79:d307:7137:79db  prefixlen 64  scopeid 0x20<link>
        inet6 2409:4042:818:ce5e:a384:5f8:91f4:380b  prefixlen 64  scopeid 0x0<global>
        ether b8:27:eb:03:05:c2  txqueuelen 1000  (Ethernet)
        RX packets 5110  bytes 3356508 (3.2 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 4480  bytes 1030865 (1006.7 KiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

```

14. From the above output, it can be noted that the AP setup indicated by "**uap0**" virtual interface is UP and RUNNING with default IP (10.0.0.1) assigned to this AP network (as we haven't given AP IP setup in step 10).


As per the above given steps and instructions, I am unable to face the issue you are facing presently.

Therefore, please try to follow the same recommended steps/instructions and provide more details which help me to debug more into the issue.

Thanks
Pankaj

**Comment by remifleurance** [comment]
I reproduced your steps, with slight adaptations to my situation, and reproduce the issue.

Here are the details:
1. Download Raspbian Buster lite (same URL, version 2020-02-13)
2. Write image to SD card (**16Go, class 10**)
3. Create wpa_supplicant.conf on /boot:
```
country=**FR**
update_config=1
ctrl_interface=/var/run/wpa_supplicant **GROUP=netdev**

network={
  scan_ssid=1
  ssid="xxxxx"
  psk="xxxxx"
}
```
4. Nothing to do here, ssh already activate in 3.
5. Boot, update, upgrade:
```
sudo apt update
sudo apt upgrade
```
6. Install git
```
sudo apt install git
```
7. Clone repo
```
git clone https://github.com/idev1/rpihotspot.git
```
8. Cd to repo
```
cd rpihotspot
```
9. Add x permissions
```
sudo chmod ug+x ./setup-network.sh
```
10. Run the script
```
sudo ./setup-network.sh --install --ap-ssid="testt          esttest" --ap-password="testtesttest" --ap-country-code="FR"
```
![image](https://user-images.githubusercontent.com/49188782/77198561-7b7cfd80-6ae7-11ea-8c6a-7bee71771336.png)

**Comment by idev1** [comment]
Hi, thanks for re-verifying the steps !
Its very strange that for you its throwing invalid country code for a valid country code as "FR" which i  still didn't replicate in my system.

But, I am trying hard to replicate this problem and once, I get any clue, I provide you more updates.

Thanks
Pankaj

**Comment by manav7** [comment]
I am getting this issue as well for country AT but since this is a optional I did not use the flag to execute with the script

**Comment by idev1** [comment]
Hi @remifleurance  and @manav7 
The Country code refusal might be related to extra spaces present in "wpa_supplicant.conf" file from where it reads the WLAN Country code. I just fixed the code to remove the extra spaces so that, the valid Country code may not be refused.

I will appreciate you if you test the branch code where the fixes are made.

Here is the link of the branch script where the fixes are available:

https://github.com/idev1/rpihotspot/blob/rpihotspot-bugfixes/setup-network.sh

Thanks
Pankaj

**Comment by ihorbond** [comment]
Hey @idev1  I got a similar error but with --ap-ip-address parameter set to "192.168.1.200" says it's invalid. Im on raspbian lite, running command with sudo bash ./setup-network ...

**Comment by idev1** [comment]
Hi @ihorbond 

Thanks for using the script and reporting an issue!

For your 1st question, i can say that in most of the WiFi router cases where we suggest Raspberry Pi to connect to a specific SSID (to access internet) and if that external WiFi router uses a local subnet starting from range "192.168.X.Y" (where, "192.168" represent chosen reserved local IP subnet range, "X" normally represents the gateway number from where it will redirect the network traffic of local subnet and lastly, "Y" represents the number which will then be assigned to each devices which request the connection to the WiFi router).

Now, if your WiFi router uses the range for e.g., "192.168.1.Y" and if you are setting up the Access Point (AP) network using the same WiFi router range (as "192.168.1.Y") then, the AP network IP addresses will conflict with your WiFi router IP address range and hence, here in this case, the script will throw an error (as you already mentioned) and abort the AP network setup.

For your second question on _not cleaning previous entries in /etc/hosts file_, yes, I do agree with you that, this piece of code should clean the entries which the script is presently unable to do but, I have noted down this point and in the future version of the script, I will provide code to resolve this observed issue. Moreover, you also said that the script is presently using a static hostname "raspberrypi" (defined in script variable ```hostName="raspberrypi"```). This is the default hostname the present downloaded Raspbian OS uses and there is no provision given in the network setup script to consider this as an input option (which can be considered in good to have feature in future release of the script) but, to handle this situation currently, you can change the said "_hostName_" entry in the script to serve your purpose of network setup.

Lastly, I will appreciate you to use the branch script which has many bug fixes related to Buster OS upgrade. Here are the steps to use the branch script.

1. Make any directory. For e.g.:
```sudo mkdir rpihotspot-bugfixes```

2. CD into the above directory:
```cd rpihotspot-bugfixes```

3. Run the following command to clone the suggested branch (in my previous post):
```sudo git clone -b rpihotspot-bugfixes https://github.com/idev1/rpihotspot.git```

4. Step 3 will create a local repo: "rpihotspot". CD into this local repo directory.
```cd rpihotspot```

5. Run the below command to verify you cloned the correct repo. The return output of this command must return * _rpihotspot-bugfixes_ which will indicate that, you correctly cloned the suggested branch.
```sudo git branch```

6. After that, you can assign the execute permission to script "_./setup-network.sh_".
```sudo chmod ug+x ./setup-network.sh```

7. Finally, you can setup the network using the install examples present in "_./readme.md_". I used the following command to successfully setup an Access Point network.
```sudo ./setup-network.sh --install --ap-ssid="*****" --ap-password="*****" --ap-country-code="IN"```


I hope, I tried my best to give the answers of your asked questions :)

Thanks
Pankaj

**Comment by ihorbond** [comment]
@idev1  thanks, you are right, 192.168.1.X is reserved that's why it wasn't working. I got it all working now, thanks. What I meant by "your script overwrites it with default "raspberrypi" is that I renamed my hostname via raspi-config -> network options but after running your script and checking /etc/hosts the 127.0.0.1 and 10.0.0.1 entries were postfixed with default "raspberrypi". I think they should follow the same hostname as in /etc/hostname

**Comment by idev1** [comment]
> @idev1 thanks, you are right, 192.168.1.X is reserved that's why it wasn't working. I got it all working now, thanks. What I meant by "your script overwrites it with default "raspberrypi" is that I renamed my hostname via raspi-config -> network options but after running your script and checking /etc/hosts the 127.0.0.1 and 10.0.0.1 entries were postfixed with default "raspberrypi". I think they should follow the same hostname as in /etc/hostname

Thanks @ihorbond for re-verifying the recommended steps!

Just to inform you that, I have considered your suggestion and now the script can clean the previously configured AP IP address entries in _/etc/hosts_ file :)

Moreover, I have also implemented a logic to read the host name entry from in-built **hostname** utility and the host name now is not defaultly hard-coded as _raspberrypi_. Below given is the code which does this.

```
# FIX: for https://github.com/idev1/rpihotspot/issues/12#issuecomment-605552834
if [ ! -z "$( hostname )" ]; then 
    hostName="$( hostname )"
fi
```

Also, all the bug fixes from branch _rpihotspot-bugfixes_ have been merged into _master_ branch and therefore, I am requesting you to use the _master_ branch to re-verify if its solving your problems those you already reported.

Thanks
Pankaj

**Comment by ihorbond** [comment]
@idev1 thanks. im trying to now turn my rpi into vpn router with cyberghost. I followed their steps described here https://support.cyberghostvpn.com/hc/en-us/articles/213270689-How-to-Set-Up-OpenVPN-on-Raspberry-Pi-Raspbian-RaspBMC- but it doesnt seem to work, I guess I need to reroute traffic via openvpn tunnel now, but I dont know how was wondering if you knew ? does the ip tables config need to be updated or do I need a bridge between one of the interfaces and a tunnel ...?

**Comment by idev1** [comment]
> @idev1 thanks. im trying to now turn my rpi into vpn router with cyberghost. I followed their steps described here https://support.cyberghostvpn.com/hc/en-us/articles/213270689-How-to-Set-Up-OpenVPN-on-Raspberry-Pi-Raspbian-RaspBMC- but it doesnt seem to work, I guess I need to reroute traffic via openvpn tunnel now, but I dont know how was wondering if you knew ? does the ip tables config need to be updated or do I need a bridge between one of the interfaces and a tunnel ...?

Hi @ihorbond 

Since the requested query is related to 3rd Party OpenVPN client and hence, i don't have any idea about. I apologize for the same.

Thanks
Pankaj

**Comment by remifleurance** [comment]
Hi Pankaj,

I cloned the new "rpihotspot-bugfixes" branch. Unfortunately this issue is still present.
Current workaround is to edit the script to replace the default value, and stop using the `-ap-country-code` parameter.

Also thank you @manav7 for your confirmation.

I will try to troubleshoot this issue by directly editing the script this week-end, but I am not very familiar with bash nor debian networking.

**Comment by idev1** [comment]
Hi @remifleurance 

After trying so many combinations, I am not sure about the reason behind ```--ap-country-code``` option not working for you. In the present _master_ branch "_setup-network.sh_" script and on line numbers: _246_ and _252_ where it sets the variable: ```apCountryCodeValid``` and if this variable becomes ```false``` then only you will have AP Country Code refusal error.

Can you please put some echo statements in the script just after line numbers: _246_ and _252_ ? and then share that information about and after which line, the variable ```apCountryCodeValid``` is becoming ```false``` so that, it will help me in understanding what going wrong with your setup/environment.

Thanks
Pankaj

**Comment by idev1** [comment]
Merged the fixed made by @remifleurance for issue #12. Hence closing issue.

**Comment by ackanir** [comment]
@idev1 Thank your for reviewing my change.
And thanks a lot for your support for keeping this tool up to date.
